File: src/parsers/opencode.ts
/**
 * parsers/opencode.ts — OpenCode sessions.
 *
 * The odd one out: not per-file but one shared SQLite DB (multi-GB, tens of
 * thousands of sessions). We open it READ-ONLY and query directly rather than
 * shell out to `opencode export` per session. Layout (ground-truthed 2026-07):
 *   session(id, parent_id, directory, title, model, agent, provider via message,
 *           time_created/updated ms, tokens_*, cost, …)
 *   message(id, session_id, time_created, data JSON{role,modelID,providerID,
 *           parentID,tokens,cost,finish,mode,path,time})
 *   part(id, message_id, session_id, time_created, data JSON{type,text,tool,
 *        callID,state,…})
 * source_path is "<db_path>#<session_id>" (no per-session file).
 *
 * Part → event mapping (one event per part):
 *   text      → assistant/user text (role from message.data)
 *   reasoning → thinking
 *   tool      → tool_call (state.input, callID) + tool_result (state.output,
 *               state.status→isError), two events sharing callID
 *   patch     → system subtype=patch (code edit)
 *   step-start/step-finish/snapshot/file → system subtype=<type> (structural markers)
 *
 * Tree A: message.data.parentID. Tree B: session.parent_id (subagent child rows,
 * titled "… (@explore subagent)"). Both preserved.
 */

import { existsSync } from "fs";
import { statSync } from "fs";
import { Database } from "bun:sqlite";
import { OPENCODE_DB_CANDIDATES } from "../paths.js";
import type { Event, EventKind, Parser, UnifiedSession } from "../types.js";
import { asDict, asNum, asStr, compact, EventBuilder, millisToEpochSec } from "../util.js";

function dbPath(): string | null {
  for (const p of OPENCODE_DB_CANDIDATES) if (existsSync(p)) return p;
  return null;
}

let _ro: Database | null = null;
function ro(): Database {
  if (_ro) return _ro;
  const p = dbPath();
  if (!p) throw new Error("opencode db not found");
  _ro = new Database(p, { readonly: true });
  return _ro;
}

interface SessRow {
  id: string;
  parent_id: string | null;
  directory: string | null;
  title: string | null;
  agent: string | null;
  model: string | null;
  time_created: number | null;
  time_updated: number | null;
  cost: number | null;
  tokens_input: number | null;
  tokens_output: number | null;
  tokens_reasoning: number | null;
  tokens_cache_read: number | null;
  tokens_cache_write: number | null;
}

export const opencodeParser: Parser = {
  agent: "opencode",

  listSessions(sinceEpochSec) {
    const p = dbPath();
    if (!p) return [];
    const d = ro();
    const rows = d
      .query<SessRow, []>(
        `SELECT id, parent_id, directory, title, agent, model, time_created, time_updated,
                cost, tokens_input, tokens_output, tokens_reasoning,
                tokens_cache_read, tokens_cache_write
         FROM session ORDER BY time_updated DESC`
      )
      .all();
    // message count + a cheap provider probe per session in one grouped query
    const counts = new Map<string, number>();
    for (const c of d
      .query<{ session_id: string; n: number }, []>(
        "SELECT session_id, COUNT(*) n FROM message GROUP BY session_id"
      )
      .all())
      counts.set(c.session_id, c.n);

    let bytes: number | null = null;
    let mtime: number | null = null;
    try {
      const st = statSync(p);
      bytes = st.size;
      mtime = Math.floor(st.mtimeMs / 1000);
    } catch {
      /* ignore */
    }

    const out: UnifiedSession[] = [];
    for (const r of rows) {
      const started = millisToEpochSec(r.time_created);
      const ended = millisToEpochSec(r.time_updated) ?? started;
      if (sinceEpochSec != null && (ended ?? 0) < sinceEpochSec) continue;
      const n = counts.get(r.id) ?? 0;
      out.push({
        id: `opencode:${r.id}`,
        agent: "opencode",
        cwd: r.directory,
        startedAt: started,
        endedAt: ended,
        title: r.title || null,
        model: r.model,
        provider: null,
        gitBranch: null,
        parentSessionId: r.parent_id ? `opencode:${r.parent_id}` : null,
        forkedFrom: null,
        turnCount: n,
        eventCount: n,
        sources: [{ path: `${p}#${r.id}`, kind: "sqlite", bytes, mtime }],
        // fingerprint: message count + updated time (grows on any new message)
        fingerprint: `${n}:${r.time_updated ?? 0}`,
        meta: {
          agent: r.agent,
          cost: r.cost,
          tokensInput: r.tokens_input,
          tokensOutput: r.tokens_output,
          tokensReasoning: r.tokens_reasoning,
          tokensCacheRead: r.tokens_cache_read,
          tokensCacheWrite: r.tokens_cache_write,
        },
      });
    }
    return out;
  },

  readEvents(nativeId): Event[] {
    if (!dbPath()) return [];
    const d = ro();
    const rows = d
      .query<
        {
          message_id: string;
          message_data: string;
          message_time: number;
          part_data: string | null;
          part_time: number | null;
        },
        [string]
      >(
        `SELECT m.id AS message_id, m.data AS message_data, m.time_created AS message_time,
                p.data AS part_data, p.time_created AS part_time
         FROM message m
         LEFT JOIN part p ON p.message_id = m.id
         WHERE m.session_id = ?
         ORDER BY m.time_created ASC, m.id ASC, p.time_created ASC, p.id ASC`
      )
      .all(nativeId);

    interface Grp {
      data: Record<string, unknown>;
      time: number;
      parts: Record<string, unknown>[];
    }
    const grouped = new Map<string, Grp>();
    const order: string[] = [];
    for (const row of rows) {
      let g = grouped.get(row.message_id);
      if (!g) {
        g = { data: safeJson(row.message_data), time: row.message_time, parts: [] };
        grouped.set(row.message_id, g);
        order.push(row.message_id);
      }
      if (row.part_data) g.parts.push(safeJson(row.part_data));
    }

    const out = new EventBuilder();
    for (const mid of order) {
      const g = grouped.get(mid)!;
      const role: EventKind = normalizeRole(asStr(g.data.role));
      const ts = millisToEpochSec(asDict(g.data.time).created) ?? millisToEpochSec(g.time);
      const parentId = asStr(g.data.parentID);
      const tok = asNum(asDict(g.data.tokens).output) ?? asNum(asDict(g.data.tokens).total);
      for (const part of g.parts) {
        const ptype = asStr(part.type) ?? "";
        if (ptype === "text") {
          out.push(role, asStr(part.text) ?? "", ts, { nativeId: mid, parentId, tokens: tok });
        } else if (ptype === "reasoning") {
          out.push("thinking", asStr(part.text) ?? "", ts, { nativeId: mid, parentId });
        } else if (ptype === "tool") {
          const name = asStr(part.tool) ?? null;
          const callId = asStr(part.callID);
          const state = asDict(part.state);
          out.push("tool_call", compact(state.input), ts, { name, callId, nativeId: mid, parentId });
          const output = state.output ?? asDict(state.metadata).output;
          if (output != null || asStr(state.status) === "error") {
            out.push("tool_result", compact(output ?? state.error ?? ""), ts, {
              name,
              callId,
              isError: asStr(state.status) === "error",
              nativeId: mid,
              parentId,
            });
          }
        } else if (ptype === "patch") {
          out.push("system", compact(part), ts, { subtype: "patch", nativeId: mid, parentId });
        } else if (ptype) {
          // step-start/step-finish/snapshot/file → structural markers
          out.push("system", compact(part), ts, { subtype: ptype, nativeId: mid, parentId });
        }
      }
    }
    return out.events;
  },
};

function safeJson(s: string | null): Record<string, unknown> {
  if (!s) return {};
  try {
    const v = JSON.parse(s);
    return v && typeof v === "object" ? v : {};
  } catch {
    return {};
  }
}

/** message.data.role → the text-event kind for that message. */
function normalizeRole(role: string | null): EventKind {
  if (role === "user") return "user";
  if (role === "system") return "system";
  return "assistant";
}

File: src/parsers/opencode.rs
use super::{Discovery, Parser, SessionCandidate};
use crate::model::{
    compact, Agent, Capture, Event, EventKind, NativeSource, ParsedSession, Session,
};
use anyhow::{Context, Result};
use rusqlite::{types::Value as SqlValue, Connection, OptionalExtension};
use serde_json::{json, Value};
use std::{
    fs,
    path::{Path, PathBuf},
};

pub struct OpenCodeParser;

type NativeSessionRow = (
    String,
    Option<String>,
    Option<String>,
    Option<String>,
    Option<String>,
    Option<String>,
    Option<i64>,
    Option<i64>,
);

fn s(v: Option<&Value>) -> Option<String> {
    v.and_then(Value::as_str).map(str::to_owned)
}
fn ms(v: Option<i64>) -> Option<i64> {
    v
}
fn parse_json(value: &str, locator: &str) -> Result<Value> {
    serde_json::from_str(value).with_context(|| format!("invalid JSON in OpenCode {locator}"))
}
fn db_path(root: &Path) -> Option<PathBuf> {
    if root.is_file() && root.extension().is_some_and(|x| x == "db") {
        Some(root.into())
    } else {
        [root.join("opencode.db"), root.join("data/opencode.db")]
            .into_iter()
            .find(|p| p.exists())
    }
}
fn parse_session(
    connection: &Connection,
    db: &Path,
    id: &str,
    _root: &Path,
    candidate: &SessionCandidate,
) -> Result<ParsedSession> {
    let (sid, parent, directory, title, agent, model, created, updated): NativeSessionRow = connection
        .query_row(
            "SELECT id,parent_id,directory,title,agent,model,time_created,time_updated FROM session WHERE id=?1",
            [id],
            |r| {
                Ok((
                    r.get(0)?,
                    r.get(1)?,
                    r.get(2)?,
                    r.get(3)?,
                    r.get(4)?,
                    r.get(5)?,
                    r.get(6)?,
                    r.get(7)?,
                ))
            },
        )?;
    let mut evs = Vec::new();
    let mut stmt=connection.prepare("SELECT m.id,m.data,m.time_created,p.data,p.time_created FROM message m LEFT JOIN part p ON p.message_id=m.id WHERE m.session_id=?1 ORDER BY m.time_created,m.id,p.time_created,p.id")?;
    let mut rows = stmt.query([id])?;
    while let Some(r) = rows.next()? {
        let mid: String = r.get(0)?;
        let md = parse_json(&r.get::<_, String>(1)?, &format!("message {mid}"))?;
        let pd: Option<String> = r.get(3)?;
        let role = md
            .get("role")
            .and_then(Value::as_str)
            .unwrap_or("assistant");
        let t = r.get::<_, Option<i64>>(4)?.or(r.get(2)?);
        if let Some(pds) = pd {
            let p = parse_json(&pds, &format!("part for message {mid}"))?;
            let typ = p.get("type").and_then(Value::as_str).unwrap_or("");
            let kind = match typ {
                "text" => {
                    if role == "user" {
                        EventKind::User
                    } else {
                        EventKind::Assistant
                    }
                }
                "reasoning" => EventKind::Thinking,
                "tool" => EventKind::ToolCall,
                "patch" | "step-start" | "step-finish" | "snapshot" | "file" => EventKind::System,
                _ => EventKind::System,
            };
            let mut e = Event::new(
                kind,
                if typ == "tool" {
                    compact(p.get("state").unwrap_or(&p))
                } else {
                    compact(p.get("text").or_else(|| p.get("output")).unwrap_or(&p))
                },
            );
            e.name = s(p.get("tool"));
            e.call_id = s(p.get("callID"));
            e.subtype = Some(typ.into());
            e.native_id = Some(mid.clone());
            e.parent_id = s(md.get("parentID"));
            e.created_at_ms = t;
            evs.push(e)
        }
    }
    let mut messages = Vec::new();
    let mut msq=connection.prepare("SELECT id,session_id,time_created,time_updated,data FROM message WHERE session_id=?1 ORDER BY time_created,id")?;
    let mut mr = msq.query([id])?;
    while let Some(r) = mr.next()? {
        let message_id = r.get::<_, String>(0)?;
        let data = parse_json(&r.get::<_, String>(4)?, &format!("message {message_id}"))?;
        messages.push(json!({"id":message_id,"session_id":r.get::<_,String>(1)?,"time_created":r.get::<_,i64>(2)?,"time_updated":r.get::<_,i64>(3)?,"data":data}));
    }
    let mut parts = Vec::new();
    let mut psq=connection.prepare("SELECT id,message_id,session_id,time_created,time_updated,data FROM part WHERE session_id=?1 ORDER BY time_created,id")?;
    let mut pr = psq.query([id])?;
    while let Some(r) = pr.next()? {
        let part_id = r.get::<_, String>(0)?;
        let data = parse_json(&r.get::<_, String>(5)?, &format!("part {part_id}"))?;
        parts.push(json!({"id":part_id,"message_id":r.get::<_,String>(1)?,"session_id":r.get::<_,String>(2)?,"time_created":r.get::<_,i64>(3)?,"time_updated":r.get::<_,i64>(4)?,"data":data}));
    }
    let envelope = json!({"format":"trace-db/opencode-session-v1","session":{"id":sid,"parent_id":parent,"directory":directory,"title":title,"agent":agent,"model":model,"time_created":created,"time_updated":updated},"message":messages,"part":parts});
    let portable_bytes = serde_json::to_vec(&envelope)?;
    let native_bytes = build_native_bundle(connection, &sid)?;
    let native_source = NativeSource {
        locator: format!("{}#{id}", db.display()),
        kind: "sqlite-session".into(),
        restore_path: format!("{sid}.db"),
        role: None,
        bytes: candidate.bytes,
        mtime_ns: candidate.mtime_ns,
        mode: candidate.mode,
        capture: Some(Capture::Bytes {
            label: sid.clone(),
            bytes: native_bytes,
        }),
    };
    let portable_source = NativeSource {
        locator: format!("{}#{id}:portable", db.display()),
        kind: "portable-json".into(),
        restore_path: format!("{sid}.json"),
        role: Some("portable-fallback".into()),
        bytes: Some(portable_bytes.len() as i64),
        mtime_ns: candidate.mtime_ns,
        mode: candidate.mode,
        capture: Some(Capture::Bytes {
            label: format!("{sid}-portable"),
            bytes: portable_bytes,
        }),
    };
    Ok(ParsedSession {
        session: Session {
            id: format!("opencode:{sid}"),
            agent: Agent::OpenCode,
            cwd: directory,
            started_at_ms: ms(created),
            ended_at_ms: ms(updated).or(created),
            title,
            model,
            provider: None,
            git_branch: None,
            parent_session_id: parent.map(|p| format!("opencode:{p}")),
            forked_from: None,
            meta: json!({"agent":agent}),
            fingerprint: format!("{}:{}", evs.len(), updated.unwrap_or_default()),
            sources: vec![native_source, portable_source],
        },
        events: evs,
    })
}

fn build_native_bundle(source: &Connection, id: &str) -> Result<Vec<u8>> {
    let temporary_directory = tempfile::tempdir()?;
    let path = temporary_directory.path().join("opencode.db");
    let connection = Connection::open(&path)?;
    let user_version: i64 = source.query_row("PRAGMA user_version", [], |row| row.get(0))?;
    connection.pragma_update(None, "user_version", user_version)?;
    clone_schema(source, &connection)?;
    connection.execute_batch("PRAGMA foreign_keys=OFF;")?;
    copy_rows(source, &connection, "migration", None, &[])?;
    let project_id: Option<String> = if table_has_column(source, "session", "project_id")? {
        source
            .query_row("SELECT project_id FROM session WHERE id=?1", [id], |row| {
                row.get(0)
            })
            .optional()?
    } else {
        None
    };
    if let Some(project_id) = project_id.as_deref() {
        copy_rows(
            source,
            &connection,
            "project",
            Some("id=?1"),
            &[SqlValue::Text(project_id.into())],
        )?;
    }
    let workspace_id: Option<String> = if table_has_column(source, "session", "workspace_id")? {
        source
            .query_row(
                "SELECT workspace_id FROM session WHERE id=?1",
                [id],
                |row| row.get(0),
            )
            .optional()?
            .flatten()
    } else {
        None
    };
    if let Some(workspace_id) = workspace_id.as_deref() {
        copy_rows(
            source,
            &connection,
            "workspace",
            Some("id=?1"),
            &[SqlValue::Text(workspace_id.into())],
        )?;
    }
    copy_rows(
        source,
        &connection,
        "session",
        Some("id=?1"),
        &[SqlValue::Text(id.into())],
    )?;
    copy_rows(
        source,
        &connection,
        "message",
        Some("session_id=?1"),
        &[SqlValue::Text(id.into())],
    )?;
    copy_rows(
        source,
        &connection,
        "part",
        Some("session_id=?1"),
        &[SqlValue::Text(id.into())],
    )?;
    connection.execute_batch(
        "CREATE TABLE IF NOT EXISTS schema_meta (key TEXT PRIMARY KEY, value TEXT NOT NULL);
         INSERT OR REPLACE INTO schema_meta(key,value) VALUES('format','opencode-native-session-v1'),('version','1');",
    )?;
    connection.execute_batch("PRAGMA foreign_keys=ON;")?;
    connection.execute_batch("VACUUM")?;
    drop(connection);
    Ok(fs::read(path)?)
}

fn quote_identifier(value: &str) -> String {
    format!("\"{}\"", value.replace('"', "\"\""))
}

fn clone_schema(source: &Connection, destination: &Connection) -> Result<()> {
    let mut tables = source.prepare(
        "SELECT sql FROM sqlite_master WHERE type='table' AND name NOT LIKE 'sqlite_%' ORDER BY name",
    )?;
    let table_sql = tables
        .query_map([], |row| row.get::<_, String>(0))?
        .collect::<rusqlite::Result<Vec<_>>>()?;
    for sql in table_sql {
        destination.execute_batch(&sql)?;
    }
    let mut objects = source.prepare(
        "SELECT sql FROM sqlite_master WHERE type IN ('index','trigger') AND sql IS NOT NULL ORDER BY type,name",
    )?;
    let object_sql = objects
        .query_map([], |row| row.get::<_, String>(0))?
        .collect::<rusqlite::Result<Vec<_>>>()?;
    for sql in object_sql {
        destination.execute_batch(&sql)?;
    }
    Ok(())
}

fn copy_rows(
    source: &Connection,
    destination: &Connection,
    table: &str,
    predicate: Option<&str>,
    values: &[SqlValue],
) -> Result<()> {
    let exists: Option<String> = source
        .query_row(
            "SELECT name FROM sqlite_master WHERE type='table' AND name=?1",
            [table],
            |row| row.get(0),
        )
        .optional()?;
    if exists.is_none() {
        return Ok(());
    }
    let table_sql = quote_identifier(table);
    let mut columns = source.prepare(&format!("PRAGMA table_info({table_sql})"))?;
    let columns = columns
        .query_map([], |row| row.get::<_, String>(1))?
        .collect::<rusqlite::Result<Vec<_>>>()?;
    let quoted_columns = columns
        .iter()
        .map(|column| quote_identifier(column))
        .collect::<Vec<_>>()
        .join(",");
    let predicate = predicate
        .map(|value| format!(" WHERE {value}"))
        .unwrap_or_default();
    let select_sql = format!("SELECT {quoted_columns} FROM {table_sql}{predicate}");
    let mut statement = source.prepare(&select_sql)?;
    let rows = statement.query_map(rusqlite::params_from_iter(values.iter()), |row| {
        (0..columns.len())
            .map(|index| row.get(index))
            .collect::<rusqlite::Result<Vec<SqlValue>>>()
    })?;
    let placeholders = (1..=columns.len())
        .map(|index| format!("?{index}"))
        .collect::<Vec<_>>()
        .join(",");
    let insert_sql = format!("INSERT INTO {table_sql} ({quoted_columns}) VALUES ({placeholders})");
    for row in rows {
        destination.execute(&insert_sql, rusqlite::params_from_iter(row?))?;
    }
    Ok(())
}

fn table_has_column(connection: &Connection, table: &str, column: &str) -> Result<bool> {
    let table_sql = quote_identifier(table);
    let mut statement = connection.prepare(&format!("PRAGMA table_info({table_sql})"))?;
    let columns = statement
        .query_map([], |row| row.get::<_, String>(1))?
        .collect::<rusqlite::Result<Vec<_>>>()?;
    Ok(columns.iter().any(|value| value == column))
}
impl Parser for OpenCodeParser {
    fn agent(&self) -> Agent {
        Agent::OpenCode
    }
    fn discover(&self, root: &Path) -> Result<Discovery> {
        let Some(db) = db_path(root) else {
            return Ok(Discovery::default());
        };
        let metadata = fs::metadata(&db)?;
        let file_candidate = SessionCandidate::file(db.clone())?;
        let c = Connection::open_with_flags(&db, rusqlite::OpenFlags::SQLITE_OPEN_READ_ONLY)?;
        let mut st =
            c.prepare("SELECT id,time_updated,time_created FROM session ORDER BY time_updated")?;
        let rows = st
            .query_map([], |row| {
                Ok((
                    row.get::<_, String>(0)?,
                    row.get::<_, Option<i64>>(1)?,
                    row.get::<_, Option<i64>>(2)?,
                ))
            })?
            .collect::<rusqlite::Result<Vec<_>>>()?;
        let mut out = Vec::new();
        for (id, updated, created) in rows {
            out.push(SessionCandidate {
                path: db.clone(),
                locator: format!("{}#{id}", db.display()),
                native_id: Some(id),
                fingerprint: format!("opencode-v1:{}", updated.or(created).unwrap_or_default()),
                updated_at_ms: updated.or(created),
                bytes: Some(metadata.len() as i64),
                mtime_ns: file_candidate.mtime_ns,
                mode: file_candidate.mode,
                parent_session_id: None,
                agent_type: None,
            });
        }
        Ok(Discovery {
            candidates: out,
            failures: Vec::new(),
        })
    }

    fn parse(&self, candidate: &SessionCandidate, root: &Path) -> Result<Option<ParsedSession>> {
        let id = candidate
            .native_id
            .as_deref()
            .ok_or_else(|| anyhow::anyhow!("OpenCode candidate missing session id"))?;
        let connection = Connection::open_with_flags(
            &candidate.path,
            rusqlite::OpenFlags::SQLITE_OPEN_READ_ONLY,
        )?;
        Ok(Some(parse_session(
            &connection,
            &candidate.path,
            id,
            root,
            candidate,
        )?))
    }

    fn parse_many(
        &self,
        candidates: &[SessionCandidate],
        root: &Path,
    ) -> Vec<(SessionCandidate, Result<Option<ParsedSession>>)> {
        let Some(first) = candidates.first() else {
            return Vec::new();
        };
        let connection = match Connection::open_with_flags(
            &first.path,
            rusqlite::OpenFlags::SQLITE_OPEN_READ_ONLY,
        ) {
            Ok(connection) => connection,
            Err(error) => {
                let error = anyhow::Error::from(error);
                return candidates
                    .iter()
                    .cloned()
                    .map(|candidate| (candidate, Err(anyhow::anyhow!("{error:#}"))))
                    .collect();
            }
        };
        candidates
            .iter()
            .cloned()
            .map(|candidate| {
                let parsed = candidate
                    .native_id
                    .as_deref()
                    .ok_or_else(|| anyhow::anyhow!("OpenCode candidate missing session id"))
                    .and_then(|id| {
                        parse_session(&connection, &candidate.path, id, root, &candidate).map(Some)
                    });
                (candidate, parsed)
            })
            .collect()
    }
}
